src/pathfinding/grid_pathfinder.jl

```diff
@@ -0,0 +1,362 @@
+export CostMetric,
+    DirectDistance,
+    Chebyshev,
+    HeightMap,
+    AStar,
+    delta_cost,
+    find_path,
+    set_target!,
+    is_stationary,
+    heightmap,
+    walkmap
+
+"""
+    Path{D}
+Alias of `MutableLinkedList{Dims{D}}`. Used to represent the path to be
+taken by an agent in a `D` dimensional [`GridSpace{D}`](@ref).
+"""
+const Path{D} = MutableLinkedList{Dims{D}}
+
+"""
+    CostMetric{D}
+An abstract type representing a metric that measures the approximate cost of travelling
+between two points in a `D` dimensional [`GridSpace{D}`](@ref).
+"""
+abstract type CostMetric{D} end
+
+struct DirectDistance{D} <: CostMetric{D}
+    direction_costs::Vector{Int}
+end
+
+Base.show(io::IO, metric::DirectDistance) = print(io, "DirectDistance")
+
+"""
+    DirectDistance
+Distance is approximated as the shortest path between the two points, provided the
+`walkable` property of [`AStar`](@ref) allows.
+Optionally provide a `Vector{Int}` that represents the cost of going from a tile to the
+neighboring tile on the `i` dimensional diagonal (default is `10√i`).
+
+If `moore_neighbors=false` in [`AStar`](@ref), only Von Neumann neighbors will be tested.
+Cost defaults to the first value of the provided vector.
+"""
+DirectDistance{D}() where {D} = DirectDistance{D}([floor(Int, 10.0 * √x) for x in 1:D])
+
+"""
+    Chebyshev
+Distance between two tiles is approximated as the Chebyshev distance (maximum of absolute
+difference in coordinates) between them.
+"""
+struct Chebyshev{D} <: CostMetric{D} end
+
+Base.show(io::IO, metric::Chebyshev) = print(io, "Chebyshev")
+
+struct HeightMap{D} <: CostMetric{D}
+    base_metric::CostMetric{D}
+    hmap::Array{Int,D}
+end
+
+Base.show(io::IO, metric::HeightMap) =
+    print(io, "HeightMap with base: $(metric.base_metric)")
+
+"""
+    HeightMap(hmap::Array{Int,D})
+    HeightMap(hmap::Array{Int,D}, ::Type{<:CostMetric})
+Distance between two positions is the sum of the shortest distance between them and the
+absolute difference in height. A heightmap of the same size as the corresponding
+[`GridSpace{D}`](@ref) is required. Distance is calculated using [`DirectDistance`](@ref)
+by default.
```

Comment:
And if not by default, what is used? The second argument? How is this obvious from the documentation string...? It isn't. Just be clear about which argument does what. Give names to the arguments like `HeightMap(hmap, cm)` and use `hmap` and `cm` in the documentation string so that it becomes clear to which argument you are referring to.